// @zetachain/faucet-cli/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_CHAINS = exports.FILE_NAME = exports.GITHUB_WHOAMI_URL = exports.GITHUB_REQUEST_TOKEN_URL = exports.GITHUB_REQUEST_CODE_URL = exports.CLIENT_ID = exports.FAUCET_VALIDATOR_URL = exports.AWS_LAMBDA_BASE = exports.LD_CLIENT_SIDE_ID = void 0;
/*
  Description: Constants used in the faucet-cli app.
  This is intended to be used out-of-the-box, therefore we shouldn't use environment variables nor shared packages.
*/
exports.LD_CLIENT_SIDE_ID = "6454092189ffbf1251467c89";
exports.AWS_LAMBDA_BASE = "https://zg8bghogy3.execute-api.us-east-1.amazonaws.com/"; // Prod env
exports.FAUCET_VALIDATOR_URL = `${exports.AWS_LAMBDA_BASE}/faucet/validator`;
exports.CLIENT_ID = "Iv1.09b6b391674ad245"; // Zeta App
exports.GITHUB_REQUEST_CODE_URL = "https://github.com/login/device/code";
exports.GITHUB_REQUEST_TOKEN_URL = "https://github.com/login/oauth/access_token";
exports.GITHUB_WHOAMI_URL = "https://api.github.com/user";
exports.FILE_NAME = "./access_token";
exports.VALID_CHAINS = ["zetachain_athens", "goerli", "bsc_testnet", "polygon_mumbai"];
